x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/amazonbedrock/AmazonBedrockService.java

```diff
@@ -0,0 +1,338 @@
+                    taskType,
+                    NAME,
+                    serviceSettings,
+                    taskSettings,
+                    secretSettings,
+                    context
+                );
+                checkProviderForTask(TaskType.TEXT_EMBEDDING, model.provider());
+                return model;
+            }
+            case COMPLETION -> {
+                var model = new AmazonBedrockChatCompletionModel(
+                    inferenceEntityId,
+                    taskType,
+                    NAME,
+                    serviceSettings,
+                    taskSettings,
+                    secretSettings,
+                    context
+                );
+                checkProviderForTask(TaskType.COMPLETION, model.provider());
+                checkChatCompletionProviderForTopKParameter(model);
+                return model;
+            }
+            default -> throw new ElasticsearchStatusException(failureMessage, RestStatus.BAD_REQUEST);
+        }
+    }
+
+    @Override
+    public TransportVersion getMinimalSupportedVersion() {
+        return ML_INFERENCE_AMAZON_BEDROCK_ADDED;
+    }
+
+    /**
+     * For text embedding models get the embedding size and
+     * update the service settings.
+     *
+     * @param model The new model
+     * @param listener The listener
+     */
+    @Override
+    public void checkModelConfig(Model model, ActionListener<Model> listener) {
+        if (model instanceof AmazonBedrockEmbeddingsModel embeddingsModel) {
+            ServiceUtils.getEmbeddingSize(
+                model,
+                this,
+                listener.delegateFailureAndWrap((l, size) -> l.onResponse(updateModelWithEmbeddingDetails(embeddingsModel, size)))
+            );
+        } else {
+            listener.onResponse(model);
+        }
+    }
+
+    private AmazonBedrockEmbeddingsModel updateModelWithEmbeddingDetails(AmazonBedrockEmbeddingsModel model, int embeddingSize) {
+        AmazonBedrockEmbeddingsServiceSettings serviceSettings = (AmazonBedrockEmbeddingsServiceSettings) model.getServiceSettings();
+        if (serviceSettings.dimensionsSetByUser()
+            && serviceSettings.dimensions() != null
+            && serviceSettings.dimensions() != embeddingSize) {
+            throw new ElasticsearchStatusException(
+                Strings.format(
+                    "The retrieved embeddings size [%s] does not match the size specified in the settings [%s]. "
+                        + "Please recreate the [%s] configuration with the correct dimensions",
+                    embeddingSize,
+                    serviceSettings.dimensions(),
+                    model.getConfigurations().getInferenceEntityId()
+                ),
+                RestStatus.BAD_REQUEST
+            );
+        }
+
+        var similarityFromModel = serviceSettings.similarity();
+        var similarityToUse = similarityFromModel == null ? SimilarityMeasure.DOT_PRODUCT : similarityFromModel;
```

Comment:
Is the default similarity always dot product and is this different for each provider?

Reply:
I think it always seems to be... this is the similarity that's used by the retriever in Elasticsearch, not the similarity on the provider itself... but I think (IIRC) - @jonathan-buttner mentioned that dot product should be the default (and Johathan - please correct me if I'm wrong ;) )

Reply:
I think it kinda depends with which similarity the model was trained/optimized for, right? But I also saw `SimilarityMeasure.DOT_PRODUCT` most of the time, so probably fair to default to that 👍

Reply:
If we know the embedding is normalised to unit length then use dot product similarity, otherwise the default should be `cosine`. 

Semantic text uses this value to create the index mappings. If the mapping is configured to use dot product similarity and the embedding is not normalised then ES will reject the embedding, that doesn't happen with cosine similarity. Using dot product confers a small efficiency as the similarity is easier to compute however, cosine is safer as it won't be rejected by ES. The default depends on the provider, so for Cohere it is dot product, HuggingFace defaults to cosine.

Reply:
I can certainly add a method to return the default similarity measure based on the provider type... for Cohere, this would be dot product, however, I can't find any information on what Amazon Titan is using under the hood. Any idea?

Reply:
BTW - the best source of information about the model I've found for Amazon Titan is here: https://docs.aws.amazon.com/bedrock/latest/userguide/titan-embedding-models.html

Reply:
(and of course, after I posted that comment...) - I just found this : https://repost.aws/questions/QUqExIWrn0SMO8mrtjSKfuFg/for-titan-embedding-model-which-distance-metric-to-use

Supposedly Amazon Titan uses cosine... I'll add a function to get the default by provider.

Reply:
If in doubt use `cosine` it's worth the small cost in performance

Reply:
All patched up - latest code now has a lookup by provider for the default.